nomad/event_endpoint.go

```diff
@@ -0,0 +1,211 @@
+	"github.com/hashicorp/nomad/nomad/stream"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+type Event struct {
+	srv *Server
+}
+
+func (e *Event) register() {
+	e.srv.streamingRpcs.Register("Event.Stream", e.stream)
+}
+
+func (e *Event) stream(conn io.ReadWriteCloser) {
+	defer conn.Close()
+
+	var args structs.EventStreamRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&args); err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	// forward to appropriate region
+	if args.Region != e.srv.config.Region {
+		err := e.forwardStreamingRPC(args.Region, "Event.Stream", args, conn)
+		if err != nil {
+			handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		}
+		return
+	}
+
+	// ACL check
+	// TODO(drew) ACL checks need to be per topic
+	// All Events        Management
+	// System Events     Management
+	// Node Events       NamespaceCapabilityReadEvents
+	// Job/Alloc Events  NamespaceCapabilityReadEvents
+	if aclObj, err := e.srv.ResolveToken(args.AuthToken); err != nil {
+		handleJsonResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil && !aclObj.IsManagement() {
+		handleJsonResultError(structs.ErrPermissionDenied, helper.Int64ToPtr(403), encoder)
+		return
+	}
+
+	// authToken is passed to the subscribe request so the event stream
+	// can handle closing a subscription if the authToken expires.
+	// If ACLs are disabled, a random token is generated and it will
+	// never be closed due to expiry.
+	authToken := args.AuthToken
+	if authToken == "" {
+		authToken = uuid.Generate()
+	}
+	subReq := &stream.SubscribeRequest{
+		Token:     authToken,
+		Topics:    args.Topics,
+		Index:     uint64(args.Index),
+		Namespace: args.Namespace,
+	}
+	publisher, err := e.srv.State().EventPublisher()
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// start subscription to publisher
+	subscription, err := publisher.Subscribe(subReq)
+	if err != nil {
+		handleJsonResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+	defer subscription.Unsubscribe()
+
+	ndJsonCh := make(chan *structs.NDJson)
+	errCh := make(chan error)
+
+	jsonStream := stream.NewNDJsonStream(ndJsonCh, 30*time.Second)
+	jsonStream.Run(ctx)
+
+	// goroutine to detect remote side closing
+	go func() {
+		if _, err := conn.Read(nil); err != nil {
+			// One end of the pipe explicitly closed, exit
+			cancel()
+			return
+		}
+		select {
+		case <-ctx.Done():
+			return
+		}
+	}()
+
+	go func() {
+		defer cancel()
+		for {
+			events, err := subscription.Next(ctx)
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-ctx.Done():
+				}
+				return
+			}
+
+			// Continue if there are no events
+			if len(events.Events) == 0 {
+				continue
+			}
```

Comment:
It is useful to send as a heartbeat?

Reply:
Depending on the request and subscription filtering, many events may end up being filtered down to 0 here, which could result in many heartbeats within a short period of time. NDJson will send one on a set interval so I think that should be enough